## This python script only merge car accident and demographic data in 2018

# Merge Outline
## 1. Accident Related Data Import
### 1.1 Accident Data Merge

## 2. Import demographic data on county level(including education, unemployment, population, and poverty)
### 2.1 Population data clean
### 2.2 Poverty data clean
### 2.3 Unemployment data clean
### 2.4 Education data clean
### 2.5 Select useful columns

## 3. Merge accident data and demographic data

In [1]:
import numpy as np
import pandas as pd

# 1. Accident Related Data Import
## Accident Related Data will be sperated into three level(case level, vehicle level, person level)
## In this project, the data will be merged in vehicle level, which means per record including one vehicle

In [2]:
### Accident level data
ACCIDENT = pd.read_csv('ACCIDENT.csv')

### Vehcicle level data
VEVENT = pd.read_csv('VEVENT.csv')
DAMAGE = pd.read_csv('DAMAGE.csv')
DISTRACT = pd.read_csv('DISTRACT.csv')
# vehicle description 
VINDECODE = pd.read_csv('VINDECODE.csv')
#vehicle circumstances which may have contributed to the crash
FACTOR = pd.read_csv('FACTOR.csv')
#action taken by the driver to avoid something or someone
MANEUVER = pd.read_csv('MANEUVER.csv')



### Person level data
PERSON = pd.read_csv('PERSON.csv')
# phicisal impairments of  driver 
DRIMPAIR = pd.read_csv('DRIMPAIR.csv')






D:\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (42) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
VEVENT

STATE  ST_CASE  EVENTNUM  VEH_NO  VEVENTNUM  VNUMBER1  AOI1  SOE  \
0           1    10001         1       1          1         1    55   63   
1           1    10001         2       1          2         1    12   14   
2           1    10001         3       1          3         1    55   69   
3           1    10001         4       1          4         1     0    1   
4           1    10001         2       2          1         1    12   14   
...       ...      ...       ...     ...        ...       ...   ...  ...   
114693     56   560100         5       1          5         1     0    1   
114694     56   560101         1       1          1         1    55   68   
114695     56   560101         2       1          2         1    11   12   
114696     56   560101         3       1          3         1     0    1   
114697     56   560101         2       2          1         1    11   12   

        VNUMBER2  AOI2  
0           5555    55  
1              2     6  
2           5555    55  
3           9999    77  
4              2     6  
...          ...   ...  
114693      9999    77  
114694      5555    55  
114695         2    11  
114696      9999    77  
114697         2    11  

[114698 rows x 10 columns]

In [4]:
DAMAGE

STATE  ST_CASE  VEH_NO  MDAREAS
0           1    10001       1        1
1           1    10001       1        2
2           1    10001       1        3
3           1    10001       1        4
4           1    10001       1        5
...       ...      ...     ...      ...
205310     56   560098       2       12
205311     56   560099       1       99
205312     56   560100       1       13
205313     56   560101       1       11
205314     56   560101       2       11

[205315 rows x 4 columns]

# 1.1 Data Merge

In [5]:
### merge data on vehicle level
vehicle_merged = pd.merge(VEVENT, DAMAGE, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')
vehicle_merged = pd.merge(vehicle_merged, DISTRACT, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')
vehicle_merged = pd.merge(vehicle_merged, VINDECODE, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')
vehicle_merged = pd.merge(vehicle_merged, FACTOR, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')
vehicle_merged = pd.merge(vehicle_merged, MANEUVER, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')
vehicle_merged = pd.merge(vehicle_merged, DRIMPAIR, on=["STATE","ST_CASE",'VEH_NO'], how= 'left')

In [6]:
vehicle_merged

STATE  ST_CASE  EVENTNUM  VEH_NO  VEVENTNUM  VNUMBER1  AOI1  SOE  \
0           1    10001         1       1          1         1    55   63   
1           1    10001         1       1          1         1    55   63   
2           1    10001         1       1          1         1    55   63   
3           1    10001         1       1          1         1    55   63   
4           1    10001         1       1          1         1    55   63   
...       ...      ...       ...     ...        ...       ...   ...  ...   
574485     56   560100         5       1          5         1     0    1   
574486     56   560101         1       1          1         1    55   68   
574487     56   560101         2       1          2         1    11   12   
574488     56   560101         3       1          3         1     0    1   
574489     56   560101         2       2          1         1    11   12   

        VNUMBER2  AOI2  ...  SUPCHRGR  SUPCHRGR_T TURBO  TURBO_T ENGVVT  \
0           5555    55  ...       NaN         NaN   NaN      NaN    NaN   
1           5555    55  ...       NaN         NaN   NaN      NaN    NaN   
2           5555    55  ...       NaN         NaN   NaN      NaN    NaN   
3           5555    55  ...       NaN         NaN   NaN      NaN    NaN   
4           5555    55  ...       NaN         NaN   NaN      NaN    NaN   
...          ...   ...  ...       ...         ...   ...      ...    ...   
574485      9999    77  ...       NaN         NaN   NaN      NaN    NaN   
574486      5555    55  ...       NaN         NaN   NaN      NaN    NaN   
574487         2    11  ...       NaN         NaN   NaN      NaN    NaN   
574488      9999    77  ...       NaN         NaN   NaN      NaN    NaN   
574489         2    11  ...         N          No     N       No      Y   

       MCYUSAGE MCYUSAGE_T MFACTOR MDRMANAV DRIMPAIR  
0           NaN        NaN     0.0     98.0      0.0  
1           NaN        NaN     0.0     98.0      0.0  
2           NaN        NaN     0.0     98.0      0.0  
3           NaN        NaN     0.0     98.0      0.0  
4           NaN        NaN     0.0     98.0      0.0  
...         ...        ...     ...      ...      ...  
574485      NaN        NaN     0.0     98.0      9.0  
574486      NaN        NaN     0.0     98.0      9.0  
574487      NaN        NaN     0.0     98.0      9.0  
574488      NaN        NaN     0.0     98.0      9.0  
574489      NaN        NaN     0.0      4.0      0.0  

[574490 rows x 115 columns]

In [7]:
# Create FIPS Code to help merge demographic data later
ACCIDENT['FIPStxt'] = ACCIDENT["STATE"].astype(str).str.zfill(2) + ACCIDENT["COUNTY"].astype(str).str.zfill(3)

In [8]:
ACCIDENT

STATE  ST_CASE  VE_TOTAL  VE_FORMS  PVH_INVL  PEDS  PERNOTMVIT  \
0          1    10001         2         1         1     0           0   
1          1    10002         1         1         0     0           0   
2          1    10003         2         2         0     0           0   
3          1    10004         1         1         0     0           0   
4          1    10005         2         2         0     0           0   
...      ...      ...       ...       ...       ...   ...         ...   
33649     56   560097         1         1         0     0           0   
33650     56   560098         2         2         0     0           0   
33651     56   560099         1         1         0     0           0   
33652     56   560100         1         1         0     0           0   
33653     56   560101         2         2         0     0           0   

       PERMVIT  PERSONS  COUNTY  ...  ARR_HOUR  ARR_MIN  HOSP_HR  HOSP_MN  \
0            1        1     121  ...         6       15       88       88   
1            2        2     127  ...         0       59       88       88   
2            2        2      21  ...        23       10       99       99   
3            2        2       3  ...        13       14       88       88   
4            2        2      73  ...         7       28       88       88   
...        ...      ...     ...  ...       ...      ...      ...      ...   
33649        1        1      13  ...         3        0        3       26   
33650        2        2      27  ...        10       58       11       43   
33651        1        1       1  ...        88       88       88       88   
33652        2        2      13  ...        99       99       99       99   
33653        2        2      13  ...        99       99       88       88   

       CF1  CF2  CF3  FATALS  DRUNK_DR  FIPStxt  
0        0    0    0       1         0    01121  
1       20    0    0       2         0    01127  
2        0    0    0       1         0    01021  
3        0    0    0       1         0    01003  
4        0    0    0       1         0    01073  
...    ...  ...  ...     ...       ...      ...  
33649    0    0    0       1         1    56013  
33650    0    0    0       1         0    56027  
33651    0    0    0       1         0    56001  
33652    0    0    0       1         1    56013  
33653    0    0    0       1         1    56013  

[33654 rows x 53 columns]

In [9]:
# accident and vehicle
accident_vehicle_merged = pd.merge(vehicle_merged, ACCIDENT, on=["STATE","ST_CASE"], how= 'left')

In [10]:
accident_vehicle_merged

STATE  ST_CASE  EVENTNUM  VEH_NO  VEVENTNUM  VNUMBER1  AOI1  SOE  \
0           1    10001         1       1          1         1    55   63   
1           1    10001         1       1          1         1    55   63   
2           1    10001         1       1          1         1    55   63   
3           1    10001         1       1          1         1    55   63   
4           1    10001         1       1          1         1    55   63   
...       ...      ...       ...     ...        ...       ...   ...  ...   
574485     56   560100         5       1          5         1     0    1   
574486     56   560101         1       1          1         1    55   68   
574487     56   560101         2       1          2         1    11   12   
574488     56   560101         3       1          3         1     0    1   
574489     56   560101         2       2          1         1    11   12   

        VNUMBER2  AOI2  ...  ARR_HOUR  ARR_MIN HOSP_HR  HOSP_MN CF1 CF2 CF3  \
0           5555    55  ...         6       15      88       88   0   0   0   
1           5555    55  ...         6       15      88       88   0   0   0   
2           5555    55  ...         6       15      88       88   0   0   0   
3           5555    55  ...         6       15      88       88   0   0   0   
4           5555    55  ...         6       15      88       88   0   0   0   
...          ...   ...  ...       ...      ...     ...      ...  ..  ..  ..   
574485      9999    77  ...        99       99      99       99   0   0   0   
574486      5555    55  ...        99       99      88       88   0   0   0   
574487         2    11  ...        99       99      88       88   0   0   0   
574488      9999    77  ...        99       99      88       88   0   0   0   
574489         2    11  ...        99       99      88       88   0   0   0   

       FATALS DRUNK_DR FIPStxt  
0           1        0   01121  
1           1        0   01121  
2           1        0   01121  
3           1        0   01121  
4           1        0   01121  
...       ...      ...     ...  
574485      1        1   56013  
574486      1        1   56013  
574487      1        1   56013  
574488      1        1   56013  
574489      1        1   56013  

[574490 rows x 166 columns]

# 2. Import demographic data

In [11]:
# demographic data including:
# 2. Population based on county
# 3. Poverty based on county
# 4. Unemployment based on county
# 5. Education based on county

population = pd.read_excel('PopulationEstimates.xlsx')
poverty = pd.read_excel('PovertyEstimates.xls')
unemployment = pd.read_excel('Unemployment.xls')
education = pd.read_excel('Education.xls')

## 2.1 Population data clean

In [12]:
# drop the useless row and reset the index for population
population.columns = population.iloc[1]
population = population.drop([0,1]).reset_index(drop=True)
population

1    FIPStxt State                         Area_Name  \
0      00000    US                     United States   
1      01000    AL                           Alabama   
2      01001    AL                    Autauga County   
3      01003    AL                    Baldwin County   
4      01005    AL                    Barbour County   
...      ...   ...                               ...   
3268   72145    PR  Vega Baja Municipio, Puerto Rico   
3269   72147    PR    Vieques Municipio, Puerto Rico   
3270   72149    PR   Villalba Municipio, Puerto Rico   
3271   72151    PR    Yabucoa Municipio, Puerto Rico   
3272   72153    PR      Yauco Municipio, Puerto Rico   

1    Rural-urban_Continuum Code_2003 Rural-urban_Continuum Code_2013  \
0                                NaN                             NaN   
1                                NaN                             NaN   
2                                  2                               2   
3                                  4                               3   
4                                  6                               6   
...                              ...                             ...   
3268                               1                               1   
3269                               7                               7   
3270                               2                               2   
3271                               1                               1   
3272                               3                               2   

1    Urban_Influence_Code_2003 Urban_Influence_Code_2013  \
0                          NaN                       NaN   
1                          NaN                       NaN   
2                            2                         2   
3                            5                         2   
4                            6                         6   
...                        ...                       ...   
3268                         1                         1   
3269                        12                        12   
3270                         2                         2   
3271                         1                         1   
3272                         2                         2   

1    Economic_typology_2015 CENSUS_2010_POP ESTIMATES_BASE_2010  ...  \
0                       NaN       308745538           308758105  ...   
1                       NaN         4779736             4780125  ...   
2                         0           54571               54597  ...   
3                         5          182265              182265  ...   
4                         3           27457               27455  ...   
...                     ...             ...                 ...  ...   
3268                    NaN           59662               59662  ...   
3269                    NaN            9301                9301  ...   
3270                    NaN           26073               26073  ...   
3271                    NaN           37941               37941  ...   
3272                    NaN           42043               41947  ...   

1    R_DOMESTIC_MIG_2019 R_NET_MIG_2011 R_NET_MIG_2012 R_NET_MIG_2013  \
0                    NaN            NaN            NaN            NaN   
1                 1.9175       0.578434        1.18631        1.52255   
2                4.84731        6.01818       -6.22612       -3.90223   
3                24.0178        16.6419        17.4886        22.7515   
4                -5.6903       0.292676       -6.89782       -8.13218   
...                  ...            ...            ...            ...   
3268                 NaN            NaN            NaN            NaN   
3269                 NaN            NaN            NaN            NaN   
3270                 NaN            NaN            NaN            NaN   
3271                 NaN            NaN            NaN            NaN   
3272                 NaN            NaN            NaN            NaN   


In [13]:
# Extracting Missing Count and Unique Count by Column
unique_count = []
for x in population.columns:
    unique_count.append([x,len(population[x].unique()),population[x].isnull().sum()])

pd.DataFrame(unique_count, columns=["Column","Unique","Missing"]).set_index("Column").T

Column   FIPStxt  State  Area_Name  Rural-urban_Continuum Code_2003  \
Unique      3273     53       2007                               10   
Missing        0      0          0                               58   

Column   Rural-urban_Continuum Code_2013  Urban_Influence_Code_2003  \
Unique                                10                         13   
Missing                               53                         58   

Column   Urban_Influence_Code_2013  Economic_typology_2015  CENSUS_2010_POP  \
Unique                          13                       7             3220   
Missing                         53                     131                0   

Column   ESTIMATES_BASE_2010  ...  R_DOMESTIC_MIG_2019  R_NET_MIG_2011  \
Unique                  3208  ...                 3186            3184   
Missing                    0  ...                   80              80   

Column   R_NET_MIG_2012  R_NET_MIG_2013  R_NET_MIG_2014  R_NET_MIG_2015  \
Unique             3181            3182            3179            3182   
Missing              80              80              80              80   

Column   R_NET_MIG_2016  R_NET_MIG_2017  R_NET_MIG_2018  R_NET_MIG_2019  
Unique             3182            3189            3182            3189  
Missing              80              80              80              80  

[2 rows x 165 columns]

## 2.2 Poverty data clean

In [14]:
# drop the useless rows and reset the index for poverty
poverty.columns = poverty.iloc[3]
poverty = poverty.drop([0,2]).reset_index(drop=True)
poverty = poverty.drop([0]).reset_index(drop=True)
poverty = poverty.drop([0]).reset_index(drop=True)
poverty

3    FIPStxt Stabr          Area_name Rural-urban_Continuum_Code_2003  \
0      00000    US      United States                             NaN   
1      01000    AL            Alabama                             NaN   
2      01001    AL     Autauga County                               2   
3      01003    AL     Baldwin County                               4   
4      01005    AL     Barbour County                               6   
...      ...   ...                ...                             ...   
3188   56037    WY  Sweetwater County                               5   
3189   56039    WY       Teton County                               7   
3190   56041    WY       Uinta County                               7   
3191   56043    WY    Washakie County                               7   
3192   56045    WY      Weston County                               7   

3    Urban_Influence_Code_2003 Rural-urban_Continuum_Code_2013  \
0                          NaN                             NaN   
1                          NaN                             NaN   
2                            2                               2   
3                            5                               3   
4                            6                               6   
...                        ...                             ...   
3188                         8                               5   
3189                         8                               7   
3190                         8                               7   
3191                        11                               7   
3192                         9                               7   

3    Urban_Influence_Code_2013 POVALL_2018 CI90LBAll_2018 CI90UBALL_2018  ...  \
0                          NaN    41852315       41619366       42085264  ...   
1                          NaN      801758         785668         817848  ...   
2                            2        7587           6334           8840  ...   
3                            2       21069          17390          24748  ...   
4                            6        6788           5662           7914  ...   
...                        ...         ...            ...            ...  ...   
3188                         8        3540           2804           4276  ...   
3189                         8        1443           1137           1749  ...   
3190                         8        2010           1585           2435  ...   
3191                        11         918            700           1136  ...   
3192                         9         777            608            946  ...   

3    CI90UB517P_2018 MEDHHINC_2018 CI90LBINC_2018 CI90UBINC_2018 POV04_2018  \
0               17.2         61937          61843          62031    3758704   
1               23.7         49881          49123          50639      73915   
2               23.9         59338          53628          65048        NaN   
3               16.9         57588          54437          60739        NaN   
4               45.9         34382          31157          37607        NaN   
...              ...           ...            ...            ...        ...   
3188            11.2         73315          67141          79489        NaN   
3189             7.7         99087          88142         110032        NaN   
3190              14         63401          56812          69990        NaN   
3191            19.3         55190          49263          61117        NaN   
3192            19.1         54319          48315          60323        NaN   

3    CI90LB04_2018 CI90UB04_2018 PCTPOV04_2018 CI90LB04P_2018 CI90UB04P_2018  
0          3714862       3802546          19.5           19.3           19.7  
1            69990         77840            26           24.6           27.4  
2              NaN           NaN           NaN            NaN            NaN  
3              NaN           NaN           NaN            NaN            NaN  
4            

In [15]:
# Extracting Missing Count and Unique Count by Column
unique_count = []
for x in poverty.columns:
    unique_count.append([x,len(poverty[x].unique()),poverty[x].isnull().sum()])

pd.DataFrame(unique_count, columns=["Column","Unique","Missing"]).set_index("Column").T

Column   FIPStxt  Stabr  Area_name  Rural-urban_Continuum_Code_2003  \
Unique      3193     52       1927                               10   
Missing        0      0          0                               57   

Column   Urban_Influence_Code_2003  Rural-urban_Continuum_Code_2013  \
Unique                          13                               10   
Missing                         57                               52   

Column   Urban_Influence_Code_2013  POVALL_2018  CI90LBAll_2018  \
Unique                          13         2803            2751   
Missing                         52            0               0   

Column   CI90UBALL_2018  ...  CI90UB517P_2018  MEDHHINC_2018  CI90LBINC_2018  \
Unique             2874  ...              508           3081            3077   
Missing               0  ...                0              0               0   

Column   CI90UBINC_2018  POV04_2018  CI90LB04_2018  CI90UB04_2018  \
Unique             3066          53             53             53   
Missing               0        3141           3141           3141   

Column   PCTPOV04_2018  CI90LB04P_2018  CI90UB04P_2018  
Unique              43              48              46  
Missing           3141            3141            3141  

[2 rows x 34 columns]

## 2.3 Unemployment data clean

In [16]:
# drop the useless row and reset the index for unemployment
unemployment = unemployment.drop([0,2]).reset_index(drop=True)
unemployment.columns = unemployment.iloc[4]
unemployment = unemployment.drop([0]).reset_index(drop=True)
unemployment = unemployment.drop([0,1]).reset_index(drop=True)
unemployment = unemployment.drop([0]).reset_index(drop=True)
unemployment = unemployment.drop([0]).reset_index(drop=True)
unemployment

4    FIPStxt Stabr                area_name Rural_urban_continuum_code_2013  \
0      00000    US            United States                             NaN   
1      01000    AL                  Alabama                             NaN   
2      01001    AL       Autauga County, AL                               2   
3      01003    AL       Baldwin County, AL                               3   
4      01005    AL       Barbour County, AL                               6   
...      ...   ...                      ...                             ...   
3270   72145    PR  Vega Baja Municipio, PR                               1   
3271   72147    PR    Vieques Municipio, PR                               7   
3272   72149    PR   Villalba Municipio, PR                               2   
3273   72151    PR    Yabucoa Municipio, PR                               1   
3274   72153    PR      Yauco Municipio, PR                               2   

4    Urban_influence_code_2013 Metro_2013 Civilian_labor_force_2000  \
0                          NaN        NaN                 142601667   
1                          NaN        NaN                   2133223   
2                            2          1                     21720   
3                            2          1                     69533   
4                            6          0                     11373   
...                        ...        ...                       ...   
3270                         1          1                     19805   
3271                        12          0                      2441   
3272                         2          1                      7636   
3273                         1          1                     10613   
3274                         2          1                     14664   

4    Employed_2000 Unemployed_2000 Unemployment_rate_2000  ...  \
0        136904680         5696987                3.99504  ...   
1          2035594           97629                    4.6  ...   
2            20846             874                      4  ...   
3            66971            2562                    3.7  ...   
4            10748             625                    5.5  ...   
...            ...             ...                    ...  ...   
3270         17607            2198                   11.1  ...   
3271          2120             321                   13.2  ...   
3272          6648             988                   12.9  ...   
3273          9005            1608                   15.2  ...   
3274         12871            1793                   12.2  ...   

4    Civilian_labor_force_2018 Employed_2018 Unemployed_2018  \
0                    161389026     155102319         6286707   
1                      2216627       2130845           85782   
2                        26196         25261             935   
3                        95233         91809            3424   
4                         8414          7987             427   
...                        ...           ...             ...   
3270                     13117         11750            1367   
3271                      2718          2269             449   
3272                      7311          6154            1157   
3273                      8762          7509            1253   
3274                      9716          8288            1428   

4    Unemployment_rate_2018 Civilian_labor_force_2019 Employed_2019  \
0                   3.89537                 163100055     157115247   
1                       3.9                   2241747       2174483   
2                       3.6                     26172         25458   
3                       3.6                     97328         94675   
4                       5.1                      8537          8213   
...                     ...                       ...           ...   
3270                   10.4                     13037         11791   
3271                   16.5                      2585          2406 

In [17]:
# Extracting Missing Count and Unique Count by Column
unique_count = []
for x in unemployment.columns:
    unique_count.append([x,len(unemployment[x].unique()),unemployment[x].isnull().sum()])

pd.DataFrame(unique_count, columns=["Column","Unique","Missing"]).set_index("Column").T

Column   FIPStxt  Stabr  area_name  Rural_urban_continuum_code_2013  \
Unique      3275     54       3274                               10   
Missing        0      0          0                               56   

Column   Urban_influence_code_2013  Metro_2013  Civilian_labor_force_2000  \
Unique                          13           3                       3142   
Missing                         56          53                          5   

Column   Employed_2000  Unemployed_2000  Unemployment_rate_2000  ...  \
Unique            3131             1756                     136  ...   
Missing              5                5                       5  ...   

Column   Civilian_labor_force_2018  Employed_2018  Unemployed_2018  \
Unique                        3139           3135             1775   
Missing                          3              3                3   

Column   Unemployment_rate_2018  Civilian_labor_force_2019  Employed_2019  \
Unique                      133                       3142           3109   
Missing                       3                          3              3   

Column   Unemployed_2019  Unemployment_rate_2019  \
Unique              1742                     125   
Missing                3                       3   

Column   Median_Household_Income_2018  \
Unique                           3082   
Missing                            82   

Column   Med_HH_Income_Percent_of_State_Total_2018  
Unique                                        3135  
Missing                                         83  

[2 rows x 88 columns]

## 2.4 Education data clean

In [18]:
education

Educational attainment for adults age 25 and older for the U.S., States, and counties, 1970-2018  \
0     Sources: U.S. Census Bureau, 1970, 1980, 1990,...                                                 
1     For definitions of rural classifications, see ...                                                 
2     This table was prepared by USDA, Economic Rese...                                                 
3                                             FIPS Code                                                 
4                                                 00000                                                 
...                                                 ...                                                 
3282                                              72145                                                 
3283                                              72147                                                 
3284                                              72149                                                 
3285                                              72151                                                 
3286                                              72153                                                 

     Unnamed: 1           Unnamed: 2                       Unnamed: 3  \
0           NaN                  NaN                              NaN   
1           NaN                  NaN                              NaN   
2           NaN                  NaN                              NaN   
3         State            Area name  2003 Rural-urban Continuum Code   
4            US        United States                              NaN   
...         ...                  ...                              ...   
3282         PR  Vega Baja Municipio                                1   
3283         PR    Vieques Municipio                                7   
3284         PR   Villalba Municipio                                2   
3285         PR    Yabucoa Municipio                                1   
3286         PR      Yauco Municipio                                3   

                     Unnamed: 4                       Unnamed: 5  \
0                           NaN                              NaN   
1                           NaN                              NaN   
2                           NaN                              NaN   
3     2003 Urban Influence Code  2013 Rural-urban Continuum Code   
4                           NaN                              NaN   
...                         ...                              ...   
3282                          1                                1   
3283                         12                                7   
3284                          2                                2   
3285                          1                                1   
3286                          2                                2   

                     Unnamed: 6                             Unnamed: 7  \
0                           NaN                                    NaN   
1                           NaN                                    NaN   
2                           NaN                                    NaN   
3     2013 Urban Influence Code  Less than a high school diploma, 1970   
4                           NaN                               52373312   
...                         ...                                    ...   
3282                          1                                    NaN   
3283                         12                                    NaN   
3284                          2                                    NaN   
3285                          1                                    NaN   
3286                          2                                    NaN   

                          Unnamed: 8                      Unnamed: 9  ...  \
0                                NaN                             NaN  ...   
1      

In [19]:
# drop the useless row and reset the index for population
education.columns = education.iloc[3]
education = education.drop([0,1]).reset_index(drop=True)
education = education.drop([0]).reset_index(drop=True)
education = education.drop([0]).reset_index(drop=True)
education['FIPStxt'] = education['FIPS Code']
education

3    FIPS Code State            Area name 2003 Rural-urban Continuum Code  \
0        00000    US        United States                             NaN   
1        01000    AL              Alabama                             NaN   
2        01001    AL       Autauga County                               2   
3        01003    AL       Baldwin County                               4   
4        01005    AL       Barbour County                               6   
...        ...   ...                  ...                             ...   
3278     72145    PR  Vega Baja Municipio                               1   
3279     72147    PR    Vieques Municipio                               7   
3280     72149    PR   Villalba Municipio                               2   
3281     72151    PR    Yabucoa Municipio                               1   
3282     72153    PR      Yauco Municipio                               3   

3    2003 Urban Influence Code 2013 Rural-urban Continuum Code  \
0                          NaN                             NaN   
1                          NaN                             NaN   
2                            2                               2   
3                            5                               3   
4                            6                               6   
...                        ...                             ...   
3278                         1                               1   
3279                        12                               7   
3280                         2                               2   
3281                         1                               1   
3282                         2                               2   

3    2013 Urban Influence Code Less than a high school diploma, 1970  \
0                          NaN                              52373312   
1                          NaN                               1062306   
2                            2                                  6611   
3                            2                                 18726   
4                            6                                  8120   
...                        ...                                   ...   
3278                         1                                   NaN   
3279                        12                                   NaN   
3280                         2                                   NaN   
3281                         1                                   NaN   
3282                         2                                   NaN   

3    High school diploma only, 1970 Some college (1-3 years), 1970  ...  \
0                          34158051                       11650730  ...   
1                            468269                         136287  ...   
2                              3757                            933  ...   
3                              8426                           2334  ...   
4                              2242                            581  ...   
...                             ...                            ...  ...   
3278                            NaN                            NaN  ...   
3279                            NaN                            NaN  ...   
3280                            NaN                            NaN  ...   
3281                            NaN                            NaN  ...   
3282                            NaN                            NaN  ...   

3    Percent of adults with a bachelor's degree or higher, 2000  \
0                                                  24.4           
1                                                    19           
2                                                    18           
3                                                  23.1           
4                                                  10.9           
...                                                 ...           
3278                                           

In [20]:
# Extracting Missing Count and Unique Count by Column
unique_count = []
for x in education.columns:
    unique_count.append([x,len(education[x].unique()),education[x].isnull().sum()])

pd.DataFrame(unique_count, columns=["Column","Unique","Missing"]).set_index("Column").T

Column   FIPS Code  State  Area name  2003 Rural-urban Continuum Code  \
Unique        3283     53       2016                               10   
Missing          0      0          0                               62   

Column   2003 Urban Influence Code  2013 Rural-urban Continuum Code  \
Unique                          13                               10   
Missing                         62                               62   

Column   2013 Urban Influence Code  Less than a high school diploma, 1970  \
Unique                          13                                   2895   
Missing                         62                                     97   

Column   High school diploma only, 1970  Some college (1-3 years), 1970  ...  \
Unique                             2695                            2019  ...   
Missing                              97                              97  ...   

Column   Percent of adults with a bachelor's degree or higher, 2000  \
Unique                                                 374            
Missing                                                 11            

Column   Less than a high school diploma, 2014-18  \
Unique                                       2714   
Missing                                        10   

Column   High school diploma only, 2014-18  \
Unique                                3021   
Missing                                 10   

Column   Some college or associate's degree, 2014-18  \
Unique                                          2971   
Missing                                           10   

Column   Bachelor's degree or higher, 2014-18  \
Unique                                   2830   
Missing                                    10   

Column   Percent of adults with less than a high school diploma, 2014-18  \
Unique                                                 328                 
Missing                                                 10                 

Column   Percent of adults with a high school diploma only, 2014-18  \
Unique                                                 370            
Missing                                                 10            

Column   Percent of adults completing some college or associate's degree, 2014-18  \
Unique                                                 294                          
Missing                                                 10                          

Column   Percent of adults with a bachelor's degree or higher, 2014-18  \
Unique                                                 433               
Missing                                                 10               

Column   FIPStxt  
Unique      3283  
Missing        0  

[2 rows x 48 columns]

## 2.5 Select useful columns
### in this stage, only the FIPS code and data in 2018 will be kept in each demographic dataframe

In [21]:
# parameters: original df(dataframe), keyword(string)
def column_selection (df, keyword):
    # FIPS_code
    FIPS_code_cols = ['FIPStxt']
    # feature
    keyword_cols = [x for x in df.columns[df.columns.str.contains(keyword)]]
    # attach 
    cols = FIPS_code_cols + keyword_cols
    # the new dataframe
    dataframe = df[cols]
    
    return dataframe

In [22]:
# use column select function to filter useful columns
education_2018 = column_selection(education,'18')
unemployment_2018 = column_selection(unemployment,'18')
population_2018 = column_selection(population,'18')
poverty_2018 = column_selection(poverty,'18')


In [23]:
population_2018

1    FIPStxt POP_ESTIMATE_2018 N_POP_CHG_2018 Births_2018 Deaths_2018  \
0      00000         326687501        1701962     3824521     2824382   
1      01000           4887681          13195       58271       53665   
2      01001             55533            143         631         518   
3      01003            217855           5334        2310        2312   
4      01005             24872           -285         263         329   
...      ...               ...            ...         ...         ...   
3268   72145             50136            NaN         NaN         NaN   
3269   72147              8362            NaN         NaN         NaN   
3270   72149             21436            NaN         NaN         NaN   
3271   72151             32254            NaN         NaN         NaN   
3272   72153             33853            NaN         NaN         NaN   

1    NATURAL_INC_2018 INTERNATIONAL_MIG_2018 DOMESTIC_MIG_2018 NET_MIG_2018  \
0             1000139                 701823                 0       701823   
1                4606                   3379              5279         8658   
2                 113                     -7                37           30   
3                  -2                     97              5224         5321   
4                 -66                     12              -231         -219   
...               ...                    ...               ...          ...   
3268              NaN                    NaN               NaN          NaN   
3269              NaN                    NaN               NaN          NaN   
3270              NaN                    NaN               NaN          NaN   
3271              NaN                    NaN               NaN          NaN   
3272              NaN                    NaN               NaN          NaN   

1    RESIDUAL_2018 GQ_ESTIMATES_2018 R_birth_2018 R_death_2018  \
0                0           8087084          NaN          NaN   
1              -69            116576      11.9381      10.9945   
2                0               455      11.3773      9.33981   
3               15              2170      10.7348      10.7441   
4                0              2813      10.5139      13.1524   
...            ...               ...          ...          ...   
3268           NaN               NaN          NaN          NaN   
3269           NaN               NaN          NaN          NaN   
3270           NaN               NaN          NaN          NaN   
3271           NaN               NaN          NaN          NaN   
3272           NaN               NaN          NaN          NaN   

1    R_NATURAL_INC_2018 R_INTERNATIONAL_MIG_2018 R_DOMESTIC_MIG_2018  \
0                   NaN                      NaN                 NaN   
1              0.943643                 0.692264             1.08152   
2               2.03745                -0.126214            0.667129   
3            -0.0092942                 0.450769             24.2764   
4              -2.63847                 0.479722            -9.23464   
...                 ...                      ...                 ...   
3268                NaN                      NaN                 NaN   
3269                NaN                      NaN                 NaN   
3270                NaN                      NaN                 NaN   
3271                NaN                      NaN                 NaN   
3272                NaN                      NaN                 NaN   

1    R_NET_MIG_2018  
0               NaN  
1           1.77379  
2          0.540916  
3           24.7272  
4          -8.75492  
...             ...  
3268            NaN  
3269            NaN  
3270            NaN  
3271            NaN  
3272            NaN  

[3273 rows x 17 columns]

# 3. Merge accident data and demographic data

In [24]:
# Merge the data based on the FIPS code
all_merged = pd.merge(accident_vehicle_merged, education_2018, on = ['FIPStxt'], how= 'left')
all_merged = pd.merge(all_merged, unemployment_2018, on = ['FIPStxt'], how= 'left')
all_merged = pd.merge(all_merged, population_2018, on = ['FIPStxt'], how= 'left')
all_merged = pd.merge(all_merged, poverty_2018, on = ['FIPStxt'], how= 'left')


In [25]:
all_merged

STATE  ST_CASE  EVENTNUM  VEH_NO  VEVENTNUM  VNUMBER1  AOI1  SOE  \
0           1    10001         1       1          1         1    55   63   
1           1    10001         1       1          1         1    55   63   
2           1    10001         1       1          1         1    55   63   
3           1    10001         1       1          1         1    55   63   
4           1    10001         1       1          1         1    55   63   
...       ...      ...       ...     ...        ...       ...   ...  ...   
574485     56   560100         5       1          5         1     0    1   
574486     56   560101         1       1          1         1    55   68   
574487     56   560101         2       1          2         1    11   12   
574488     56   560101         3       1          3         1     0    1   
574489     56   560101         2       2          1         1    11   12   

        VNUMBER2  AOI2  ...  CI90UB517P_2018  MEDHHINC_2018 CI90LBINC_2018  \
0           5555    55  ...             30.1          41630          38789   
1           5555    55  ...             30.1          41630          38789   
2           5555    55  ...             30.1          41630          38789   
3           5555    55  ...             30.1          41630          38789   
4           5555    55  ...             30.1          41630          38789   
...          ...   ...  ...              ...            ...            ...   
574485      9999    77  ...               24          51204          47219   
574486      5555    55  ...               24          51204          47219   
574487         2    11  ...               24          51204          47219   
574488      9999    77  ...               24          51204          47219   
574489         2    11  ...               24          51204          47219   

        CI90UBINC_2018 POV04_2018 CI90LB04_2018 CI90UB04_2018 PCTPOV04_2018  \
0                44471        NaN           NaN           NaN           NaN   
1                44471        NaN           NaN           NaN           NaN   
2                44471        NaN           NaN           NaN           NaN   
3                44471        NaN           NaN           NaN           NaN   
4                44471        NaN           NaN           NaN           NaN   
...                ...        ...           ...           ...           ...   
574485           55189        NaN           NaN           NaN           NaN   
574486           55189        NaN           NaN           NaN           NaN   
574487           55189        NaN           NaN           NaN           NaN   
574488           55189        NaN           NaN           NaN           NaN   
574489           55189        NaN           NaN           NaN           NaN   

       CI90LB04P_2018 CI90UB04P_2018  
0                 NaN            NaN  
1                 NaN            NaN  
2                 NaN            NaN  
3                 NaN            NaN  
4                 NaN            NaN  
...               ...            ...  
574485            NaN            NaN  
574486            NaN            NaN  
574487            NaN            NaN  
574488            NaN            NaN  
574489            NaN            NaN  

[574490 rows x 223 columns]

In [26]:
# Extracting Missing Count and Unique Count by Column
unique_count = []
for x in all_merged.columns:
    unique_count.append([x,len(all_merged[x].unique()),all_merged[x].isnull().sum()])

pd.DataFrame(unique_count, columns=["Column","Unique","Missing"]).set_index("Column").T

Column   STATE  ST_CASE  EVENTNUM  VEH_NO  VEVENTNUM  VNUMBER1  AOI1  SOE  \
Unique      51    33654        28      27         21        27    27   69   
Missing      0        0         0       0          0         0     0    0   

Column   VNUMBER2  AOI2  ...  CI90UB517P_2018  MEDHHINC_2018  CI90LBINC_2018  \
Unique         23    27  ...              496           2753            2759   
Missing         0     0  ...               91             91              91   

Column   CI90UBINC_2018  POV04_2018  CI90LB04_2018  CI90UB04_2018  \
Unique             2748           1              1              1   
Missing              91      574490         574490         574490   

Column   PCTPOV04_2018  CI90LB04P_2018  CI90UB04P_2018  
Unique               1               1               1  
Missing         574490          574490          574490  

[2 rows x 223 columns]

### the all_merged data including 223 features and some missing values

In [32]:
# output the dataframe to CSV file
all_merged.to_csv('all_merged.csv', index=False)